Source/BoundaryConditions/WarpX_PEC.H

```diff
@@ -0,0 +1,172 @@
+     *        The number or depth of guard cells updated is equal to the shape factor of
+     *        particles in each dimension.
+     *
+     * \param[in] icomp        component of the Efield being updated
+                               (0=x, 1=y, 2=z in 3D) and (0=x, 1=z in 2D)
+     * \param[in] dom_lo       index value of the lower domain boundary (cell-centered)
+     * \param[in] dom_hi       index value of the higher domain boundary (cell-centered)
+     * \param[in] ijk_vec      indices along the x(i), y(j), z(k) of Efield Array4
+     * \param[in] Efield       field data to be updated if (ijk) is at the boundary or a guard cell
+     * \param[in] stag         staggering of the field data being updated.
+     * \param[in] fbndry_lo    Field boundary type at the lower boundaries
+     * \param[in] fbndry_hi    Field boundary type at the upper boundaries
+     */
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    void SetTangentialEfield (const int icomp, const amrex::IntVect & dom_lo,
+                                const amrex::IntVect &dom_hi,
+                                const amrex::IntVect &ijk_vec,
+                                amrex::Array4<amrex::Real> const& Efield,
+                                const amrex::IntVect& stag,
+                                amrex::GpuArray<int, 3> const& fbndry_lo,
+                                amrex::GpuArray<int, 3> const& fbndry_hi )
+    {
+        // Fields in guard cells set equal and opposite to valid cells
+        amrex::IntVect ijk_valid = ijk_vec;
+        for (int idim = 1; idim < AMREX_SPACEDIM; ++idim) {
+            // index of dimension
+            const int dir = ( ( (icomp+idim)<AMREX_SPACEDIM)? (icomp+idim) : ( (icomp+idim) - AMREX_SPACEDIM));
+            // loop over sides , iside = 0 (lo), iside = 1 (hi)
+            for (int iside = 0; iside < 2; ++iside) {
+                const int bndry_cell = ( (iside == 0)
+                                     ? (ijk_vec[dir] == dom_lo[dir])
+                                     : (ijk_vec[dir] == (dom_hi[dir] + stag[dir]) ) );
+                const int is_tangent_to_PEC = ( ( iside == 0 )
+                                            ? is_boundary_PEC(fbndry_lo, dir)
+                                            : is_boundary_PEC(fbndry_hi, dir) );
+                // if boundary is PEC, guard cells must be equal and opposite to valid cells
+                if ( bndry_cell == 1 && is_tangent_to_PEC==1) {
+                    Efield(ijk_vec) = 0.;
+                }
+                const int guard_cell = ( (iside == 0)
+                                     ? (ijk_vec[dir] < dom_lo[dir])
+                                     : (ijk_vec[dir] > (dom_hi[dir] + stag[dir]) ) );
+                if ( guard_cell == 1 && is_tangent_to_PEC == 1) {
+                    // guard cell outside the domain by "ig" number of cells
+                    const int ig = ( (iside == 0)
```

Comment:
Two comments:
Why is the stag[dir] not included here?
The overall logic could be simplified by calculating ig at the start, then you know it's a boundary cell if ig == 0 and a guard cell if ig > 0.

Reply:
oh that is to be fixed! Thank you so much for catching this!